### ISSUE TITLE
Move to ESM

### ISSUE THREAD

**Issue by LitoMore** [vadimdemedes/import-jsx#15]
It blocks https://github.com/vadimdemedes/create-ink-app/issues/13

**Comment by sindresorhus** [comment]
This will need to use [loader hooks](https://nodejs.org/api/esm.html#esm_loaders).

**Comment by yamalight** [comment]
It seems like creating a wrapper around dynamic import would also work, i.e.:
```js
export const importJsx = async (moduleId, options) => {
  // ....
  const modulePath = resolveFrom(dirname(callerPath()), moduleId);

  // oversimplified loading without in-mem caching / etc
  // read to string
  const source = (await readFile(modulePath)).toString();
  // compile + cache to file using babel
  const result = cache({
    modulePath,
    options,
    source,
    version
  });

  // encode JS
  const encodedJs = encodeURIComponent(result);
  // create data URL
  const dataUri = 'data:text/javascript;charset=utf-8,' + encodedJs;
  // load that URL as a module
  const m = await import(dataUri);

  // ...
  return m;
};
```

Usage would be similar to dynamic import itself, e.g.:
```js
import importJsx from 'import-jsx';

const MyReactComponent = await importJsx('./component.jsx');

// do something with component ...
```

I think that _should_ work. Not sure how different that'd be to using loader hooks though 🤔

**Comment by LitoMore** [comment]
@yamalight All [`require` related things](https://github.com/vadimdemedes/import-jsx/blob/master/index.js#L23-L29) need to be refactored.

**Comment by yamalight** [comment]
@LitoMore I could try to do that if you're open to dynamic import approach.
I think in-mem caching would need to be implemented separately since we'll be loading modules from strings (so default ESM cache wouldn't work as expected).

**Comment by ajstrand** [comment]
Is there any progress on this? Would it be possible to create a workaround to ESM like a custom babel transform for all the files?

**Comment by vadimdemedes** [comment]
I've tried implementing @yamalight's idea, but it doesn't work for importing nested JSX files, since modules imported via `data:` URLs can't import anything else → https://nodejs.org/dist/latest-v18.x/docs/api/esm.html#data-imports:

> Resolving [relative specifiers](https://nodejs.org/dist/latest-v18.x/docs/api/esm.html#terminology) does not work because data: is not a [special scheme](https://url.spec.whatwg.org/#special-scheme). For example, attempting to load ./foo from data:text/javascript,import "./foo"; fails to resolve because there is no concept of relative resolution for data: URLs.

For example:

**root.js**
```js
import importJsx from 'import-jsx';

await importJsx('./a.js');
```

**a.js**
```js
import importJsx from 'import-jsx';

await importJsx('./b.js');
```

**b.js**
```js
// Hello world
```

In this scenario, it's not possible to import b.js. Unless anyone has any ideas how to solve this, unfortunately this doesn't look like a viable replacement to current `import-jsx` functionality.

**Comment by vadimdemedes** [comment]
This leaves loader hooks as the only option and there are already packages that built this:

```
$ node --loader=ts-node/esm root.js
$ node --loader=@esbuild-kit/esm-loader root.js
```

**Comment by vadimdemedes** [comment]
Submitted https://github.com/vadimdemedes/import-jsx/pull/18, which converts `import-jsx` to ESM and uses loader hooks.

**Comment by vadimdemedes** [comment]
Released in [5.0.0](https://github.com/vadimdemedes/import-jsx/releases/tag/v5.0.0).
